src/gstools/field/pgs.py

```diff
@@ -0,0 +1,173 @@
+
+.. currentmodule:: gstools.field.pgs
+
+The following classes are provided
+
+.. autosummary::
+   :toctree:
+
+   PGS
+"""
+
+# pylint: disable=C0103
+import numpy as np
+
+# very clunky way of supporting both np 1.x and 2.x exceptions
+try:
+    np.AxisError = np.exceptions.AxisError
+except AttributeError:
+    ...
+
+
+class PGS:
+    """A class to generate plurigaussian field simulations (PGS).
+
+    See e.g. [Ricketts2023]_ and [Armstrong2011]_ for more details.
+
+    Parameters
+    ----------
+    dim : :class:`int`
+        dimension of the field
+    fields : :class:`list` or :class:`numpy.ndarray`
+        For `dim > 1` a list of spatial random fields (SRFs), with
+        `len(fields) == dim`. For `dim == 1`, the SRF can be directly given,
+        instead of a list. This class supports structured and unstructured meshes.
+        All fields must have the same shapes.
+
+    Notes
+    -----
+    Using plurigaussian fields for conditioning fields is still a beta feature.
+
+    References
+    ----------
+    .. [Ricketts2023] Ricketts, E.J., Freeman, B.L., Cleall, P.J. et al.
+        A Statistical Finite Element Method Integrating a Plurigaussian Random
+        Field Generator for Multi-scale Modelling of Solute Transport in
+        Concrete. Transp Porous Med 148, 95–121 (2023)
+        https://doi.org/10.1007/s11242-023-01930-8
+    .. [Armstrong2011] Armstrong, Margaret, et al.
+        Plurigaussian simulations in geosciences.
+        Springer Science & Business Media, 2011.
+        https://doi.org/10.1007/978-3-642-19607-2
+    """
+
+    def __init__(self, dim, fields):
+        # hard to test for 1d case
+        if dim > 1:
+            if dim != len(fields):
+                raise ValueError(
+                    "PGS: Mismatch between dim. and no. of fields."
+                )
+        for d in range(1, dim):
+            if not fields[0].shape == fields[d].shape:
+                raise ValueError("PGS: Not all fields have the same shape.")
+        self._dim = dim
+        self._Zs = np.array(fields)
+        self._L = None
+        self._pos_l = None
+        try:
+            self._mapping = np.stack(self._Zs, axis=1)
+        except np.AxisError:
+            # if dim==1, `fields` is prob. a raw field & not a 1-tuple or
+            # equivalent
+            if self._dim == 1:
+                self._Zs = np.array([self._Zs])
+                self._mapping = np.stack(self._Zs, axis=1)
+            else:
+                raise
+
+    def __call__(self, L):
+        """Generate the plurigaussian field.
+
+        Parameters
+        ----------
+        L : :class:`numpy.ndarray`
+            A `dim` dimensional structured field, whose values are mapped to the PGS.
+            It does not have to have the same shape as the `fields`, as the indices are
+            automatically scaled.
+        Returns
+        -------
+        pgs : :class:`numpy.ndarray`
+            the plurigaussian field
+        """
```

Comment:
Here the same applies. I wouldn't use `L` for this argument. It should be something concise and understandable.